src/modsim2/similarity/metrics/kde.py

```diff
@@ -0,0 +1,324 @@
+        def func(*args):
+            return np.power(
+                np.exp(estimator_A.score([[*args]]))
+                - np.exp(estimator_B.score([[*args]])),
+                2,
+            )
+
+        # Perform integration
+        distance = KDE._kde_distance(
+            num_dimensions=num_dimensions,
+            func=func,
+            integration_kwargs=integration_kwargs,
+        )
+        distance = np.sqrt(distance)
+
+        return distance, distance
+
+    @staticmethod
+    def kl(
+        num_dimensions: int,
+        estimator_A: KernelDensity,
+        estimator_B: KernelDensity,
+        integration_kwargs: dict,
+    ) -> tuple[float, float]:
+        """
+        Calculates the KL divergence between two probability density estimatiors for a
+        given number of dimensions using integration
+
+        Args:
+            num_dimensions : the number of dimensions that are to be integrated over
+            estimator_A: the density estimator that has been generated for a dataset A
+            estimator_B: the density estimator that has been generated for a dataset B
+
+        Returns:
+            divergence: the KL divergence
+        """
+        # the function to be integrated
+        def func(A, B, *args):
+            return np.exp(A.score(np.array([[*args]]))) * (
+                np.log(
+                    (np.exp(A.score([[*args]])))
+                    / (np.exp(B.score(np.array([[*args]]))))
+                )
+            )
+
+        def funcAB(*args):
+            return func(estimator_A, estimator_B, *args)
+
+        def funcBA(*args):
+            return func(estimator_B, estimator_A, *args)
+
+        # perform integration
+        divergenceAB = KDE._kde_distance(
+            num_dimensions=num_dimensions,
+            func=funcAB,
+            integration_kwargs=integration_kwargs,
+        )
+        divergenceBA = KDE._kde_distance(
+            num_dimensions=num_dimensions,
+            func=funcBA,
+            integration_kwargs=integration_kwargs,
+        )
+
+        return divergenceAB, divergenceBA
+
+    @staticmethod
+    def total_variation(
+        num_dimensions: int,
+        estimator_A: KernelDensity,
+        estimator_B: KernelDensity,
+        integration_kwargs: dict,
+    ) -> tuple[float, float]:
+        """
+        Calculates the total variation between two probability density estimatiors for
+        a given number of dimensions using integration
+
+        Args:
+            num_dimensions : the number of dimensions that are to be integrated over
+            estimator_A: the density estimator that has been generated for a dataset A
+            estimator_B: the density estimator that has been generated for a dataset B
+
+        Returns:
+            distance: the total variation distance
+        """
+        # the function to be integrated
+        def func(*args):
+            return np.abs(estimator_A.score([[*args]] - estimator_B.score([[*args]])))
```

Comment:
Error: estimators return the log-likelihood so should be applying the exponential function (as per the other integration functions)